pkg/reconciler/taskrun/validate_resources.go

```diff
@@ -122,6 +122,59 @@ func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params
 		return fmt.Errorf("param types don't match the user-specified type: %s", wrongTypeParamNames)
 	}
 
+	return ValidateObjectParameters(paramSpecs, params)
+}
+
+// ValidateObjectParameters checks if all required keys of object type params are provided in taskrun params.
+// To use this function externally, make sure `paramSpecs` and `params` have
+// been checked against general missing/extra params and param type alignment,
+// which been done in `validateParams` in this module.
+func ValidateObjectParameters(paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param) error {
+	neededKeys := make(map[string][]string)
+	providedKeys := make(map[string][]string)
+
+	// collect needed keys for object parameters
+	for _, spec := range paramSpecs {
+		if spec.Type == v1beta1.ParamTypeObject {
+			for key := range spec.Properties {
+				neededKeys[spec.Name] = append(neededKeys[spec.Name], key)
+			}
+		}
+	}
+
+	// collect provided keys for object parameters
+	for _, p := range params {
+		if p.Value.Type == v1beta1.ParamTypeObject {
+			for key := range p.Value.ObjectVal {
+				providedKeys[p.Name] = append(providedKeys[p.Name], key)
+			}
+		}
+	}
+
+	return ValidateObjectKeysNeededAgainstProvided(neededKeys, providedKeys)
+}
+
+// ValidateObjectKeysNeededAgainstProvided checks if objects have missing keys in its provider (either taskrun value or result value)
+// To use this function at result side, make sure `providedObjectParamKeys` and `neededObjectParamKeys` have
+// been checked against missing/extra object which been done in `validateParams` in this module.
+func ValidateObjectKeysNeededAgainstProvided(neededObjectKeys, providedObjectKeys map[string][]string) error {
+	missings := []string{}
+	for p, keys := range providedObjectKeys {
+		if _, ok := neededObjectKeys[p]; !ok {
+			// Ignore any missing objects - this happens when extra objects were
+			// passed that aren't being used.
+			continue
+		}
+		missedKeys := list.DiffLeft(neededObjectKeys[p], keys)
+		if len(missedKeys) != 0 {
+			missings = append(missings, p)
+		}
+	}
+
+	if len(missings) != 0 {
+		return fmt.Errorf("object provider misses providing the keys required for %v", missings)
+	}
```

Comment:
Because we want to give users more informative error message. For example, you have object param `a`, `b` .. `d`and all of them have keys missing in provider (from taskrun or result). If we did what you just said and directly return error, what will happen is that once key missing is detected for `a`, error will return and tell users "hey, parameter a has missing keys, please fix it", and users realize it and change it and apply their yaml. In the second round, they received error message again "hey param b has missing keys, please fix again.", so on and so forth. You can see, if many object params have missing keys, users need to apply the yaml many many times in order to get what params have missing keys.

I don't think this is user-friendly error handling. So in this implementation, we collect all params once which have missing keys, and tell users  "hey, these object param [a, b, c, d] have missing keys, please fix them".

Reply:
Yes this makes sense! I think it is a trade-off, like whether we want to save more user's time like this. And throw err in advance will not have significant improvement in our case

Reply:
> And throw err in advance will not have significant improvement in our case
exactly